Bitpacking mode info

arjenpdevries opened the pull request:
Version of #15622 without (I think) merge conflicts, which was a continuation of #15544 

This implements the suggested serialize/deserialize, in line with the suggestions by @Tishj (and building on his initials steps).

More detailed info in #15544

arjenpdevries committed:
This Xmas, I wanted BitpackingMode info in `pragma_storage_info` but Santa did not bring it;
so I tried myself ;-)

This pull request will do:

- Create a BitpackingSegmentState that overrides GetSegmentInfo to return mode info
- Add BitpackingInitSegment that returns segment-info (used in ColumnData)
- Add BitpackingInitSegment / init_segment to the compressionfunction

(I also renamed BitpackingCompressState to BitpackingCompressionState for
consistency with other compression code.)

The approach implemented seems to work correctly and passes tests locally.

However, the new information is only shown when you first close the database and reopen it.
If you only use `CHECKPOINT`, `init_segment` is not called and the modes not reported.
I am stuck where to fix this; it seems that this would be a step related to `ConvertToPersistent`
that I overlook but I do not understand what I need to do to make it work.

I have tested using the following SQL (but because of this incomplete behaviour not yet added as a test):

```
pragma force_compression = 'BitPacking';
CREATE or replace TABLE test (a INTEGER, b INTEGER);
insert into test values (10,12), (11,12), (12,11), (NULL,NULL);
insert into test values (10,12), (33,33), (33,33), (10,12);
select segment_id, row_group_id, block_id, block_offset, compression, segment_info FROM pragma_storage_info('test') order by segment_id, row_group_id  asc;
checkpoint;
select segment_id, row_group_id, block_id, block_offset, compression, segment_info FROM pragma_storage_info('test') order by segment_id, row_group_id  asc;
```

After closing the CLI and restarting it on the same database, and then doing
```
select segment_id, row_group_id, block_id, block_offset, compression, segment_info FROM pragma_storage_info('test') order by segment_id, row_group_id  asc;
```

the `pragma_storage_info` reports `segment_info` that includes the bitpacking mode.

Hoping for some DuckDB compression guru advice, A.

```
D select segment_id, row_group_id, block_id, block_offset, compression, segment_info FROM pragma_storage_info('test') order by segment_id, row_group_id  asc;
┌────────────┬──────────────┬──────────┬──────────────┬──────────────┬──────────────┐
│ segment_id │ row_group_id │ block_id │ block_offset │ compression  │ segment_info │
│   int64    │    int64     │  int64   │    int64     │   varchar    │   varchar    │
├────────────┼──────────────┼──────────┼──────────────┼──────────────┼──────────────┤
│          0 │            0 │        1 │            0 │ BitPacking   │ Mode: for    │
│          0 │            0 │        1 │           48 │ Uncompressed │              │
│          0 │            0 │        1 │          304 │ BitPacking   │ Mode: for    │
│          0 │            0 │        1 │          352 │ Uncompressed │              │
└────────────┴──────────────┴──────────┴──────────────┴──────────────┴──────────────┘
```

arjenpdevries committed:
ran formatter

arjenpdevries committed:
Does not need to include itself

arjenpdevries committed:
Implemented Serialize/Deserialize for the Bitmode Histogram

arjenpdevries committed:
Minor

arjenpdevries committed:
Merge branch 'main' into bitpacking-mode-info

arjenpdevries committed:
Merge branch 'bitpacking-mode-info' of github.com:informagi/duckdb into bitpacking-mode-info

arjenpdevries committed:
Serializer class forward def needed

arjenpdevries committed:
Use msp instead of unordered_map so order in info remains fixed (easier processing, simpler tests).

arjenpdevries committed:
Satisfy tidy?

Tishj commented on src/storage/compression/bitpacking.cpp:
```diff
@@ -949,16 +973,70 @@ void BitpackingSkip(ColumnSegment &segment, ColumnScanState &state, idx_t skip_c
+	return std::move(result);
+}
+
+void CleanupState(ColumnSegment &segment) {
```
I don't think we need this, cleanup state is necessary for state that holds lifetimes outside of the scope of the segment, such as additional blocks that it has created - which is what the `UncompressedStringSegmentState` uses it for.

UncompressedString can create additional blocks to hold "overflow" strings, which are bigger than a certain threshold (I think it's 4096 by default)
When the column segment that these additional blocks are a part of gets dropped, we need to inform the block manager that these additional blocks can now also be safely reused.

arjenpdevries replied:
Of course, should have realized. Will remove.

arjenpdevries replied:
Just out of curiosity - we now write extra information, could it happen that this requires an extra block (in the case where everything was filled completely with data and I want to write out the bitmode histogram) or does this extra serialization data get written to a separate place in storage?

Tishj reviewed (changes requested):
Thanks, this looks like it'll work from reading the code 👍 
I do however want to see at least one test making sure this is working properly

Looking at:
```c++
static constexpr const idx_t BITPACKING_METADATA_GROUP_SIZE = STANDARD_VECTOR_SIZE > 512 ? STANDARD_VECTOR_SIZE : 2048;
```
We should be deciding per 2048 values which mode to use, so given 10000 tuples to compress, this should result in a total of 5 groups
(also your test should probably use `require vector_size 2048` because the vector size will impact the group size)

arjenpdevries committed:
Adapted function names for (hopefully successful) amalgation, and removed use of class template as T is not needed.

arjenpdevries committed:
Added tests for the new functionality

arjenpdevries committed:
removed unnecessary cleanup implementation

Tishj reviewed (approved):
Thanks, looks great! 👍

Mytherin reviewed (commented):
Thanks for the PR!

Unless I'm misunderstanding the PR - this seems to store more information in the database only to provide the segment information during the `storage_info` callback. The new serialize/deserialize logic will also hurt forwards compatibility of database files. 

Given this information exists elsewhere already in the blocks, perhaps we can instead introduce a separate `get_segment_info` callback to the compression method so that we don't need to introduce the new compressed segment state?

Tishj commented:
The PR adds `segment_info` for the bitpacked column segments

```sql
query I
SELECT
	segment_info
FROM
	pragma_storage_info('test_bp')
WHERE segment_type NOT IN ('VALIDITY')
----
CONSTANT: 9, DELTA_FOR: 1
```

I'm not sure about the serialization / forwards compatibility complications, but this sounds like a welcome change for observability ?
But now that I think about it, I guess this does get complicated by the fact that old versions don't produce this information, and we need to safely handle this

I also feel like Logging this information during checkpointing would be a nicer way to provide this observability, waiting patiently on #15119 🙏

Mytherin commented:
The change itself is welcome for sure - my point is that this information is already stored in the (compressed) data itself. We should be able to just read it from the blocks. That then (1) also works with older database files, (2) does not require storing additional data in the database file, increasing the size, and (3) does not break the ability of older DuckDB versions from reading this compression due to the introduction of new information they do not understand.

arjenpdevries commented:
Thanks Mark!

My initial approach used only init-segment, but didn't work at the moment of checkpointing; in discussion we switched to this design. 

I guess I can alternatively scan all segments in init-segment to recreate this information upon the column being loaded from disk. Have to look into the exact mechanics but will try. 

(It may imply a longer load time?)

Tishj commented:
```c++
	void LoadNextGroup() {
		D_ASSERT(bitpacking_metadata_ptr > handle.Ptr() &&
		         bitpacking_metadata_ptr < handle.Ptr() + current_segment.GetBlockManager().GetBlockSize());
		current_group_offset = 0;
		current_group = DecodeMeta(reinterpret_cast<bitpacking_metadata_encoded_t *>(bitpacking_metadata_ptr));

		bitpacking_metadata_ptr -= sizeof(bitpacking_metadata_encoded_t);
```

All the group metadata is neatly packed together, so scanning just that portion should be cheap and efficient

Mytherin commented:
> Thanks Mark!
> 
> My initial approach used only init-segment, but didn't work at the moment of checkpointing; in discussion we switched to this design.
> 
> I guess I can alternatively scan all segments in init-segment to recreate this information upon the column being loaded from disk. Have to look into the exact mechanics but will try.
> 
> (It may imply a longer load time?)

`GetSegmentInfo` is only called very rarely (when the `storage_info ` function is called) - and is not very performance sensitive. Ideally the code introduced in this PR is only ever run when the user calls `storage_info` - and does not impact any other (much more common) code paths (i.e. does not slow down compression, load, etc).

Tishj commented:
> Thanks Mark!
> 
> My initial approach used only init-segment, but didn't work at the moment of checkpointing; in discussion we switched to this design.
> 
> I guess I can alternatively scan all segments in init-segment to recreate this information upon the column being loaded from disk. Have to look into the exact mechanics but will try.
> 
> (It may imply a longer load time?)

Yes, so you were mostly on the right track to do what Mark is suggesting, but the place it was being done was wrong.
You were doing this in InitSegment, which receives both fresh and existing blocks
For an existing block this information is available, but for a fresh one it doesn't exist yet

arjenpdevries commented:
Hooking into `InitSegment` is still an option, but I'm afraid it will get called whenever the column is loaded from disk into memory - and not only when pragma-storage-info would be requested. Would that be a problem?

arjenpdevries commented:
PS: for the fresh blocks, the current implementation (yours ;-)) does not need the serialize/deserialize. But we need to trigger the same process when a column is loaded from disk. I do not know where to best put that trigger - if we do this in `InitSegment` we will always run it, but when `pragma-storage-info` requests this information, there is no obvious hook into the segment initiating this "go initialize that metadata now please".

Maybe an alternative is to trigger scanning from the call to `GetSegmentState` (defined in `column_segment.hpp`) by calling `InitSegment` there (and therefore moving the definition of `GetSegmentState` to `column_segment.cpp`, where also other calls to `init_segment` live). Is that a better idea?

Mytherin commented:
> Given this information exists elsewhere already in the blocks, perhaps we can instead introduce a separate get_segment_info callback to the compression method so that we don't need to introduce the new compressed segment state?

I would go with this approach ^

Tishj commented:
@arjenpdevries this is the diff, without cleaning up the existing serialize/deserialize code

```patch
diff --git a/src/include/duckdb/function/compression_function.hpp b/src/include/duckdb/function/compression_function.hpp
index d11beaa72b..cda6e9538c 100644
--- a/src/include/duckdb/function/compression_function.hpp
+++ b/src/include/duckdb/function/compression_function.hpp
@@ -203,6 +203,12 @@ typedef unique_ptr<ColumnSegmentState> (*compression_deserialize_state_t)(Deseri
 //! Function prototype for cleaning up the segment state when the column data is dropped
 typedef void (*compression_cleanup_state_t)(ColumnSegment &segment);
 
+//===--------------------------------------------------------------------===//
+// GetSegmentInfo (optional)
+//===--------------------------------------------------------------------===//
+//! Function prototype for retrieving segment information straight from the column segment
+typedef string (*compression_get_segment_info_t)(ColumnSegment &segment);
+
 class CompressionFunction {
 public:
 	CompressionFunction(CompressionType type, PhysicalType data_type, compression_init_analyze_t init_analyze,
@@ -297,6 +303,13 @@ public:
 	compression_deserialize_state_t deserialize_state;
 	//! Cleanup the segment state (optional)
 	compression_cleanup_state_t cleanup_state;
+
+	// Get Segment Info
+	//! This is only necessary if you want to convey more information about the segment in the 'pragma_storage_info'
+	//! result
+
+	//! Get stringified segment information directly from reading the column segment
+	compression_get_segment_info_t get_segment_info = nullptr;
 };
 
 //! The set of compression functions
diff --git a/src/storage/compression/bitpacking.cpp b/src/storage/compression/bitpacking.cpp
index 8f6a593dbf..65132e9ae6 100644
--- a/src/storage/compression/bitpacking.cpp
+++ b/src/storage/compression/bitpacking.cpp
@@ -1012,6 +1012,27 @@ unique_ptr<ColumnSegmentState> BitpackingDeserializeState(Deserializer &deserial
 	return std::move(result);
 }
 
+template <class T>
+string BitpackingGetSegmentInfo(ColumnSegment &segment) {
+	map<BitpackingMode, idx_t> counts;
+	auto tuple_count = segment.count.load();
+	BitpackingScanState<T> scan_state(segment);
+	for (idx_t i = 0; i < tuple_count; i += BITPACKING_METADATA_GROUP_SIZE) {
+		if (i) {
+			scan_state.LoadNextGroup();
+		}
+		counts[scan_state.current_group.mode]++;
+	}
+
+	vector<string> result;
+	for (auto &it : counts) {
+		auto &mode = it.first;
+		auto &count = it.second;
+		result.push_back(StringUtil::Format("%s: %d", EnumUtil::ToString(mode), count));
+	}
+	return StringUtil::Join(result, ", ");
+}
+
 //===--------------------------------------------------------------------===//
 // Get Function
 //===--------------------------------------------------------------------===//
@@ -1025,6 +1046,7 @@ CompressionFunction GetBitpackingFunction(PhysicalType data_type) {
 	bitpacking.init_segment = BitpackingInitSegment;
 	bitpacking.serialize_state = BitpackingSerializeState;
 	bitpacking.deserialize_state = BitpackingDeserializeState;
+	bitpacking.get_segment_info = BitpackingGetSegmentInfo<T>;
 	return bitpacking;
 }
 
diff --git a/src/storage/table/column_data.cpp b/src/storage/table/column_data.cpp
index 53e7c54597..d8d560fc13 100644
--- a/src/storage/table/column_data.cpp
+++ b/src/storage/table/column_data.cpp
@@ -881,10 +881,15 @@ void ColumnData::GetColumnSegmentInfo(idx_t row_group_index, vector<idx_t> col_p
 		} else {
 			column_info.persistent = false;
 		}
-		auto segment_state = segment->GetSegmentState();
-		if (segment_state) {
-			column_info.segment_info = segment_state->GetSegmentInfo();
-			column_info.additional_blocks = segment_state->GetAdditionalBlocks();
+		auto &compression_function = segment->GetCompressionFunction();
+		if (compression_function.get_segment_info) {
+			column_info.segment_info = compression_function.get_segment_info(*segment);
+		} else {
+			auto segment_state = segment->GetSegmentState();
+			if (segment_state) {
+				column_info.segment_info = segment_state->GetSegmentInfo();
+				column_info.additional_blocks = segment_state->GetAdditionalBlocks();
+			}
 		}
 		result.emplace_back(column_info);
 
```

arjenpdevries committed:
Merge branch 'main' into bitpacking-mode-info

arjenpdevries commented:
Hey @Tishj in the patch of `column_data.cpp` we now either call `get_segment_info` or the segment's `GetSegmentState`'s `GetSegmentInfo`. 

Wouldn't it be better to adapt the second part?

It now returns by default an empty string through its definition in `CompressedSegmentState` class defined in `include/duckdb/function/compression_function.hpp` so in that case it could simply be dropped.

The default is only overridden in `include/duckdb/storage/checkpoint/string_checkpoint_state.hpp` - which could also be moved into a new `get_segment_info` for `StringUncompressed` (in `storage/compression/string_uncompressed.cpp`).

If we modify it like that, it is consistent for all (the two) compression methods that use it at the moment, and we can remove the `GetSegmentInfo` call in `CompressedSegmentState` - I think it was only put there as a workaround because this `get_segment_info` function was missing.

What do you think?
Shall I make that change too?

arjenpdevries committed:
Strategy suggested by Mark: add separate get_segment_info callback.
Thanks Thijs for providing a patch.

arjenpdevries committed:
Define the new get_segment_info for Bitpacking;
undo the serialization/deserialization of the cached bitmode hists.

arjenpdevries committed:
Call the new get_segment_info callback, instead of the old
GetSegmentInfo method in CompressedSegmentState.

Tishj commented:
Hey, I will be busy until the release, and we're entering feature freeze shortly / we have entered feature freeze (with some exceptions) so I will not have time to look at this or help with it for the coming weeks

arjenpdevries commented:
Okay of course, but I do think it's done - so maybe in time for feature freeze? 

(I think the CI is reporting unrelated issues from main?)

arjenpdevries committed:
Merge branch 'main' into bitpacking-mode-info
Attempt to fix CI (I think it is the Boost::filesystem in iceberg that Carlopi removed that will fix things)

arjenpdevries commented:
I think it is ready!

Mytherin commented on src/include/duckdb/storage/compression/bitpacking.hpp:
```diff
@@ -17,4 +18,17 @@ enum class BitpackingMode : uint8_t { INVALID, AUTO, CONSTANT, CONSTANT_DELTA, D
 
+class Serializer;
+
+struct SerializedBitpackingSegmentState : public ColumnSegmentState {
```
Can this be removed?

Mytherin commented on src/storage/compression/bitpacking.cpp:
```diff
@@ -949,16 +949,42 @@ void BitpackingSkip(ColumnSegment &segment, ColumnScanState &state, idx_t skip_c
+// GetSegmentInfo
+//===--------------------------------------------------------------------===//
+template <class T>
+string BitpackingGetSegmentInfo(ColumnSegment &segment) {
```
Maybe it makes sense to turn this return type into a `InsertionOrderPreservingMap<string>` similar to  `PhysicalOperator::ParamsToString`?

Tishj replied:
I guess `segment_info` becomes a `MAP(VARCHAR, VARCHAR)` then?
I do like that better for future-proofing yea

arjenpdevries replied:
Then `GetSegmentInfo` (for `String Uncompressed`) needs to change too. Alternatively, keep `segment_info` a string now but serialize it from the map?

Mytherin reviewed (commented):
Thanks for the PR! Looks a lot cleaner now - @Tishj perhaps you want to have a look as well. Some comments:

arjenpdevries committed:
Oops my mistake!

arjenpdevries committed:
Merge branch 'bitpacking-mode-info' of github.com:informagi/duckdb into bitpacking-mode-info

arjenpdevries committed:
rewrite if-then-else

arjenpdevries committed:
include no longer needed

arjenpdevries committed:
Modify get_segment_info return type

Turn return type into a InsertionOrderPreservingMap<string> similar to PhysicalOperator::ParamsToString (following suggestion by Mark).
Keep segment_info column of type VARCHAR to limit changes.

arjenpdevries commented:
So followed the suggestion to use the map but kept segment_info a column of VARCHAR for now. LMK if you'd rather see that changed. Otherwise seems good to go.

Tishj reviewed (approved):
Thanks, I think it's fine to keep it as VARCHAR for now 👍

arjenpdevries commented:
(only removed two includes, it should pass tests as before)

Mytherin commented:
Looks like they were not unnecessary :)

arjenpdevries commented:
OK don't believe the LSP without checking twice...

arjenpdevries commented:
_Will not touch it any more!_ 😀

Mytherin commented:
Thanks!
